hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/ClientNonSmartListenerService.java

```diff
@@ -22,101 +22,126 @@
         EventHandler handler = registrationKey.getHandler();
-        ClientMessage request = registrationKey.getRequest();
-
         handler.beforeListenerRegister();
+        ClientMessage request = registrationKey.getCodec().encodeAddRequest(false);
         ClientInvocation invocation = new ClientInvocation(client, request);
         invocation.setEventHandler(handler);
-        try {
-            ClientInvocationFuture future = invocation.invoke();
-            String registrationId = registrationKey.getCodec().decodeAddResponse(future.get());
-            handler.onListenerRegister();
-            Address address = future.getInvocation().getSendConnection().getRemoteEndpoint();
-            ClientEventRegistration registration = new ClientEventRegistration(registrationId,
-                    request.getCorrelationId(), address, registrationKey.getCodec());
-            registrations.put(registrationKey, registration);
-        } catch (Exception e) {
-            //if invocation cannot be done that means connection is broken and there is no need to add listener
-            EmptyStatement.ignore(e);
-        }
+
+        ClientInvocationFuture future = invocation.invoke();
+        String registrationId = registrationKey.getCodec().decodeAddResponse(future.get());
+        handler.onListenerRegister();
+        Address address = future.getInvocation().getSendConnection().getRemoteEndpoint();
+        return new ClientEventRegistration(registrationId,
+                request.getCorrelationId(), address, registrationKey.getCodec());
+
     }
 
     @Override
     public boolean deregisterListener(String userRegistrationId) {
-        ClientEventRegistration registration = registrations.remove(new ClientRegistrationKey(userRegistrationId));
-        if (registration == null) {
-            return false;
-        }
-        removeEventHandler(registration.getCallId());
-        ClientMessage request = registration.getCodec().encodeRemoveRequest(registration.getServerRegistrationId());
-        try {
-            Future future = new ClientInvocation(client, request, registration.getSubscriber()).invoke();
-            future.get();
-        } catch (Exception e) {
-            //if invocation cannot be done that means connection is broken and listener is already removed
-            EmptyStatement.ignore(e);
+        synchronized (listenerRegLock) {
+            ClientRegistrationKey key = new ClientRegistrationKey(userRegistrationId);
+            ClientEventRegistration registration = registrations.get(key);
+
+            if (registration == null) {
+                return false;
+            }
+
+            ClientMessage request = registration.getCodec().encodeRemoveRequest(registration.getServerRegistrationId());
+            try {
+                Future future = new ClientInvocation(client, request).invoke();
+                future.get();
+                removeEventHandler(registration.getCallId());
```

Comment:
previously we removed the event handler before the invocation, now it's done after. Is this intentional?